### DISCUSSION TITLE
S3 within ELFinder configuration

### DISCUSSION THREAD

**Discussion by hhhc** [Laravel-Backpack/community-forum#395] [Q&A (Help)]
Hello,

in order to prepare the deployment on Laravel Vapor, I need to configure  the filemanager add-on to be used with S3.
Information about the proper configuration is very difficult to find and I have tried hours to get it working.
Last things I tried could at least connect to S3 but it is an ugly hack and "Your configuration files are not serializable" :-(

In config/elfinder.php I did:

```
    'roots' => array(
        array(
            'driver'        => 'LocalFileSystem',           // driver for accessing file system (REQUIRED)
            'path'          => 'uploads',  // path to files (REQUIRED)
        ),
        array(
            'driver'        => 'LocalFileSystem',
            'path'          => 'places',
            'plugin' => array(
                'AutoResize' => array(
                    'enable'         => true,       // On / off
                    'maxWidth'       => 320,       // Max width
                    'maxHeight'      => 320,       // max height
                    'quality'        => 85,         // JPEG image save quality
                    'preserveExif'   => false,      // Preserve EXIF data (Imagick only)
                    'forceEffect'    => false,      // For change quality or make progressive JPEG of small images
                ),
            ),
        ),
        [
            'driver' => 'Flysystem',
            'URL' => 'http://mydomain.com/content',
            'alias' => 'S3-storage',
            'filesystem' => new \League\Flysystem\Filesystem(
                new \League\Flysystem\AwsS3V3\AwsS3V3Adapter(
                    new \Aws\S3\S3Client(
                        Array ( "driver" => "s3",
                            "key" => env('AWS_FILEMANAGER_ACCESS_KEY_ID'),
                            "secret" => env('AWS_FILEMANAGER_SECRET_ACCESS_KEY'),
                            "region" => env('AWS_FILEMANAGER_REGION'),
                            "bucket" => env('AWS_FILEMANAGER_BUCKET'),
                            "url" => env('AWS_FILEMANAGER_BUCKET'),
                            "version" => "latest",
                            "credentials" => Array (
                                "key" => env('AWS_FILEMANAGER_ACCESS_KEY_ID'),
                                "secret" => env('AWS_FILEMANAGER_SECRET_ACCESS_KEY'),
                            )
                        ) 
                    ),
                    env('AWS_FILEMANAGER_BUCKET')
                )
            )
        ],
    ),
```

I would be very thankful if anybody can share a working configuration. 
Thanks!

**Comment by pxpm** [comment]
Hello @hhhc I've never tried that myself. Maybe https://github.com/santhapa/elfinder-aws-s3 could help you ?

**Reply by hhhc** [reply]
Hello @pxpm, thanks for the prompt reply. The suggested answer is actually where I got a lot of inspiration. Unfortunately, all those solutions I could find in the Internet are standalone solutions, not really integrated into Laravel / Backpack.
So, my above solution actually works if you don't want to cache the config. However, as Laravel Vapor always does it this solution no longer works.
So, in case you have another idea it would be much appreciated.

Thanks

**Reply by pxpm** [reply]
Hey @hhhc indeed, if you have closures and stuff like that in your config files you will lose the ability to cache them since they could not be serialized anymore. 

If you need to create a new instance of the class it seems that what you need is to create a new class that extends that one with your custom configurations and pass it as a string to the package ?

**Comment by hhhc** [comment]
Hi @pxpm, my mistake was that I tried to do everything in $roots not in $disks.
Now I have a working solution (see also https://github.com/barryvdh/elfinder-flysystem-driver/issues/89).
```
    'disks' => [
        's3' => [
			'URL' => env('AWS_FILEMANAGER_URL'),
			'alias' => 'Uploads',
		],
    ],

```

However, when I use the "browse" field type in conjunction with the filemanager there are some open questions.

I am using this CRUD field 
```
        CRUD::field('image')
            ->label("Image")
            ->type("browse")
            ->size(12);
```

but it does not put the full URL in the DB but the path:
“Uploads/2023/singapore/molnar-balint-k-ecibbfku0-unsplash.jpg”

If the alias is left blank, the value stored in the DB is “s3/2023/singapore/molnar-balint-k-ecibbfku0-unsplash.jpg”.

Both are incorrect when running this via a manually managed S3 bucket (in a Laravel Vapor environment) as the Cloudfront Url for this bucket would need to be prepended. In the filemanager, the URL is correct.

<img width="452" alt="Bildschirm­foto 2023-02-10 um 09 38 49" src="https://user-images.githubusercontent.com/961087/218043698-0891d328-b9f1-4c02-a7f5-c3a5f920f3ec.png">

Any suggestions on how to do that?

**Reply by pxpm** [reply]
Maybe you need to configure more stuff in your elfinder config file or in the disk itself ? 

https://github.com/Studio-42/elFinder/wiki/Connector-configuration-options-2.1#root-options

I am thinking maybe `path` or similar option for the roots?? Also there is `url` in filesystem to configure similar stuff.  

Cheers

**Comment by jcastroa87** [comment]
Hello,

We haven’t heard back from you in a while, so we’re closing this issue for now. If you still need assistance, please don’t hesitate to reopen the issue or create a new one.

Thank you, and take care!

Cheers.
